website/pages/docs/connect/observability/ui-visualization.mdx

```diff
@@ -0,0 +1,501 @@
+     *         {label: "SR", desc: "...", value: "99%"},
+     *         ...
+     *       ],
+     *       ...
+     *     }
+     *   }
+     */
+    downstreamRecentSummaryStats: function(serviceDC, namespace, serviceName, options) {}
+  }
+
+  // Register the provider with Consul for use. This example would be usable by
+  // configuring the agent with `ui_config.metrics_provider = "example-provider".
+  window.consul.registerMetricsProvider("example-provider", provider)
+
+}());
+```
+
+Additionally, the built in [Prometheus
+provider code](https://github.com/hashicorp/consul/blob/master/ui/packages/consul-ui/vendor/metrics-providers/prometheus.js)
+can be used as a reference.
+
+### Configuring the Agent With a Custom Metrics Provider.
+
+In the example below, we configure the Consul agent to use a metrics provider
+named `example-provider`, which is defined in
+`/usr/local/bin/example-metrics-provider.js`. The name `example-provider` must
+have been specified in the call to `consul.registerMetricsProvider` as in the
+code listing in the last section.
+
+```hcl
+ui_config {
+  enabled = true
+  metrics_provider = "example-provider"
+  metrics_provider_files = ["/usr/local/bin/example-metrics-provider.js"]
+  metrics_provider_options_json = <<-EOT
+    {
+      "foo": "bar"
+    }
+  EOT
+}
+```
+
+More than one javascript file may be specified in
+[`metrics_provider_files`](/docs/agent/options#ui_config_metrics_provider_files)
+and all we be served allowing flexibility if needed to include dependencies.
+Only one metrics provider can be configured and used at one time.
+
+The
+[`metrics_provider_options_json`](/docs/agent/options#ui_config_metrics_provider_options_json)
+field is an optional literal JSON object which is passed to the provider's
+`init` method at startup time. This allows configuring arbitrary parameters for
+the provider in config rather than hard coding them into the provider itself to
+make providers more reusable.
+
+The provider may fetch metrics directly from another source although in this
+case the agent will probably need to serve the correct CORS headers to prevent
+browsers from blocking these requests. These may be configured with
+[`http_config.response_headers`](http://localhost:3000/docs/agent/options#response_headers).
+
+Alternatively, the provider may choose to use the [built-in metrics
+proxy](#metrics-proxy) to avoid cross domain issues or to inject additional
+authorization headers without requiring each UI user to be separately
+authenticated to the metrics backend. To do so, requests should be made to
+`/v1/internal/ui/metrics-proxy/<path>`. All components in the `<path>` will be
+appended to the configured
+[`base_url`](/docs/agent/options#ui_config_metrics_proxy_base_url) in the
+agent's metrics proxy configuration, and the request forwarded. The response
+will be returned without any modification to be interpreted by the provider and
+converted into the format as described in the interface above.
```

Comment:
```suggestion
authenticated to the metrics backend.

A fetch-like function is provided to the metrics provider javascript during
`init` as `options.fetch`. This is a thin fetch wrapper that prefixes any url
with the url of Consul's metrics proxy endpoint and adds your current Consul
ACL token to the request headers. Otherwise it functions like the browsers
native fetch and will forward your request on to the metrics backend. The
response will be returned without any modification to be interpreted by the
provider and converted into the format as described in the interface above.
```